internal/inband/actions.go

```diff
@@ -0,0 +1,220 @@
+	"github.com/metal-toolbox/flasher/internal/runner"
+	rtypes "github.com/metal-toolbox/rivets/types"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+
+	imodel "github.com/metal-toolbox/ironlib/model"
+)
+
+var (
+	errCompose = errors.New("error in composing steps for firmware install")
+)
+
+const (
+	// transition types implemented and defined further below
+	powerOffServer         model.StepName = "powerOffServer"
+	powerCycleServer       model.StepName = "powerCycleServer"
+	checkInstalledFirmware model.StepName = "checkInstalledFirmware"
+	downloadFirmware       model.StepName = "downloadFirmware"
+	installFirmware        model.StepName = "installFirmware"
+	pollInstallStatus      model.StepName = "pollInstallStatus"
+)
+
+const (
+	PreInstall  model.StepGroup = "PreInstall"
+	PostInstall model.StepGroup = "PostInstall"
+	Install     model.StepGroup = "Install"
+	PowerState  model.StepGroup = "PowerState"
+)
+
+type ActionHandler struct {
+	handler *handler
+}
+
+func (i *ActionHandler) identifyComponent(ctx context.Context, component string, models []string, deviceQueryor device.InbandQueryor) (*rtypes.Component, error) {
+	var components rtypes.Components
+
+	if len(i.handler.task.Server.Components) > 0 {
+		components = rtypes.Components(i.handler.task.Server.Components)
+	} else {
+		deviceCommon, err := deviceQueryor.Inventory(ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		components, err = model.NewComponentConverter().CommonDeviceToComponents(deviceCommon)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	found := components.ByNameModel(component, models)
+	if found == nil {
+		// nolint:goerr113 // its clearer to define this error here
+		errComponentMatch := fmt.Errorf(
+			"unable to identify component '%s' from inventory for given models: %s",
+			component,
+			strings.Join(models, ","),
+		)
+
+		return nil, errComponentMatch
+	}
+
+	return found, nil
+}
+
+func (i *ActionHandler) ComposeAction(ctx context.Context, actionCtx *runner.ActionHandlerContext) (*model.Action, error) {
+	var deviceQueryor device.InbandQueryor
+	if actionCtx.DeviceQueryor == nil {
+		deviceQueryor = NewDeviceQueryor(actionCtx.Logger)
+	} else {
+		deviceQueryor = actionCtx.DeviceQueryor.(device.InbandQueryor)
+	}
+
+	i.handler = initHandler(actionCtx, deviceQueryor)
+
+	component, err := i.identifyComponent(ctx, actionCtx.Firmware.Component, actionCtx.Firmware.Models, deviceQueryor)
+	if err != nil {
+		return nil, errors.Wrap(ErrComponentNotFound, err.Error())
+	}
+
+	i.handler.logger.WithFields(logrus.Fields{
+		"component": actionCtx.Firmware.Component,
+		"model":     component.Model,
+		"current":   component.Firmware.Installed,
+	}).Info("target component identified for firmware install")
+
+	required, err := deviceQueryor.FirmwareInstallRequirements(
+		ctx,
+		actionCtx.Firmware.Component,
+		actionCtx.Firmware.Vendor,
+		component.Model,
+	)
+	if err != nil {
+		// not a fatal error
```

Comment:
How do we know its not a fatal error? Could it also be something else? We shouldnt assume that the only error here is there are no install requirements.

Reply:
Its only a fatal error if no update utility was available in ironlib, thanks for pointing this out